File: temporal_profile.py
from .base import TemporalProfile
import numpy as np

# %% ############################# ####
#### Options for temporal profiles ####
#### ############################# ####


class GaussianTemporal(TemporalProfile):
    def __init__(self, tau):
        """
        Gaussian temporal profile:
            envelope(t) = exp(-2 t^2 / tau^2)
        tau : float
            RMS pulse duration [fs]
        """
        self.tau = tau

    def __call__(self, t):
        t = np.asarray(t, dtype=float)
        return np.exp(-2.0 * t**2 / self.tau**2)

File: csrc/src/temporal_profile.c
#include <math.h>
#include "temporal_profile.h"

/* --------------------- Gaussian temporal --------------------- */

static double gaussian_temporal_eval(const TemporalProfile *base, double t_fs)
{
    const GaussianTemporal *self = (const GaussianTemporal *)base;
    const double tau = self->tau_fs;
    return exp(-2.0 * (t_fs * t_fs) / (tau * tau));
}

static void gaussian_temporal_destroy(TemporalProfile *base)
{
    (void)base;
}

static const TemporalProfileVTable GAUSSIAN_TEMPORAL_VTABLE = {
    .eval = gaussian_temporal_eval,
    .destroy = gaussian_temporal_destroy};

void GaussianTemporal_init(GaussianTemporal *g, double tau_fs)
{
    g->base.vt = &GAUSSIAN_TEMPORAL_VTABLE;
    g->tau_fs = tau_fs;
}
